### ISSUE TITLE
SQLite: Use WAL mode

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#3303]
If we ever have problems with SQLite performance, we might want to look into WAL mode:

> There are advantages and disadvantages to using WAL instead of a rollback journal. Advantages include:
> - WAL is significantly faster in most scenarios.
> ...

https://sqlite.org/wal.html
https://github.com/simonw/til/blob/main/sqlite/enabling-wal-mode.md

**Comment by internalfx** [comment]
I currently run under WAL mode which is recommended for performance.

Also it band-aids the problem prisma has with concurrency #2955 because writers won't block readers.

This at least allows me to use prisma for reads and a different library for writes.

**Comment by Wulfheart** [comment]
Do I understand correctly that Prisma doesn't support WAL as migration yet?

**Comment by buraks** [comment]
Hi, since WAL is not supported Prisma becomes unusable in our project. We have two applications sharing a single SQLite database. Even though the applications are not doing a lot of writes, they often get blocked by each other. To mitigate a little bit, we have set `?connection_limit=1` on both applications, however, it's not solving the problem entirely.

It would be amazing if Prisma supported WAL, do you have an estimated timeline for its support?

**Comment by janpio** [comment]
No, we do not publish timelines but only sometimes communicate when we put issues on our public roadmap (https://pris.ly/roadmap). This issues is currently not on that roadmap and is not high priority to end up there right now. (Your comment of course is a vote to change that, but we have a lot of other issues that are more pressing right now)

**Comment by jacob-ebey** [comment]
It would be awesome to support WAL mode to be able to use Prisma with https://litestream.io

**Comment by jlarmstrongiv** [comment]
@jacob-ebey have you found any workarounds?  I’m trying to integrate https://litestream.io/ too

**Comment by Brendonovich** [comment]
+1 for this, being able to use https://litestream.io/ would be brilliant

**Comment by jan-kn** [comment]
I am using Prisma with SQLite in WAL mode without any issues. According to the SQLite docs, WAL mode is persistent. Just run `PRAGMA journal_mode=WAL;` once or add it to a migration.

> Unlike the other journaling modes, PRAGMA journal_mode=WAL is persistent. If a process sets WAL mode, then closes and reopens the database, the database will come back in WAL mode.
> [...]
> The persistence of WAL mode means that applications can be converted to using SQLite in WAL mode without making any changes to the application itself. One has merely to run "PRAGMA journal_mode=WAL;" on the database file(s) using the command-line shell or other utility, then restart the application.
>
> -- [https://sqlite.org/wal.html](https://sqlite.org/wal.html)

As long as I don't run any migrations [Litestream](https://litestream.io/) also works just fine. It's only Prisma Migrate that complains about the `_litestream_lock` and `_litestream_seq` tables. Sometimes I also get a "database is locked" error during migrations while Litestream is running in the background:

```
Error: SQLite database error
database is locked
   0: migration_core::commands::apply_migrations::Applying migration
           with migration_name="20220825190018_add_post"
             at migration-engine/core/src/commands/apply_migrations.rs:91
   1: migration_core::state::ApplyMigrations
             at migration-engine/core/src/state.rs:185
```

**Comment by Indeedornot** [comment]
As I understand it there is no option as of yet to instantiate a new database in WAL mode?

**Comment by exKAZUu** [comment]
FYI: we run the following script before starting a app depending prisma (for Litestream):

```sh
node -e '
const { PrismaClient } = require("@prisma/client");
new PrismaClient().$queryRaw`PRAGMA journal_mode = WAL;`
  .catch((error) => { console.log("Failed due to:", error); process.exit(1); });
'
```

**Comment by Fire-Hound** [comment]
Thank you for writing the solution after solving it on your own ❤️

**Comment by jdkdev** [comment]
I would love WAL mode as well.

I am also using Prisma with Sqlite.  +1!!

**Comment by matjaeck** [comment]
SQLite is used *a lot* but not many talk about it as it has something "unprofessional" sticking to it outside of embedding it into other products. But this is changing, as investments/products like Fly's LiteFS show. Not everyone needs distributed SQLite, but everyone needs replication for backup and restore when using SQLite in production, which happens more than you'd think. Do I understand that Prisma views this as a low priority feature assuming they are targeting the "serious app developers"? Yes. Do I think they miss out on a trend? Yes. Would I want them to focus on this? No.

It's a very specific implementation and that does not fit well to Prismas approach of abstracting SQL and DB-specifics "away". You'd have to know exactly how writes are performed and that goes against the philosophy.

>  He [the founder] envisioned databases becoming as straightforward as creating a page in Notion, rather than a complex, fragile structure requiring constant attention.

So IMO people who want to explore Litestream and possibilities of SQLite at scale are better suited with simpler ORM's or just query builders which put control and responsibility in their hands and don't add a massive abstraction suite that is mostly a limitation for such specific use cases.

**Comment by jkomyno** [comment]
Something to take into account: `PRAGMA journal_mode=WAL` (or [other custom journal modes](https://www.sqlite.org/pragma.html#pragma_journal_mode), for that matter) are not supported in [Cloudflare D1](https://developers.cloudflare.com/d1/).

**Comment by reustle** [comment]
Based on the helpful answers by @jan-kn and others above, here is the solution I've used. I figured it would be useful to add some clarity to the steps.

Create an empty migration
```bash
prisma migrate dev --create-only --name sqlite-wal-mode
```

Update the new `migration.sql` file to include
```sql
-- Enable WAL mode in SQLite
PRAGMA journal_mode=WAL;
```

Run the new migration
```bash
prisma migrate dev
```

You should now see a `*-wal` file in the same directory as your database.

*(Please let me know if I missed something)*
